modules/core/src/main/java/org/apache/ignite/internal/visor/metric/VisorMetricTask.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.ignite.spi.metric.LongMetric;
+import org.apache.ignite.spi.metric.Metric;
+import org.apache.ignite.spi.metric.ObjectMetric;
+import org.apache.ignite.spi.metric.ReadOnlyMetricRegistry;
+import org.jetbrains.annotations.Nullable;
+
+import static org.apache.ignite.internal.processors.metric.impl.MetricUtils.SEPARATOR;
+import static org.apache.ignite.spi.metric.jmx.MetricRegistryMBean.searchHistogram;
+
+/** Reperesents visor task for obtaining metric values. */
+@GridInternal
+@GridVisorManagementTask
+public class VisorMetricTask extends VisorOneNodeTask<VisorMetricTaskArg, Map<String, ?>> {
+    /** */
+    private static final long serialVersionUID = 0L;
+
+    /** {@inheritDoc} */
+    @Override protected VisorJob<VisorMetricTaskArg, Map<String, ?>> job(VisorMetricTaskArg arg) {
+        return new VisorMetricJob(arg, false);
+    }
+
+    /** */
+    private static class VisorMetricJob extends VisorJob<VisorMetricTaskArg, Map<String, ?>> {
+        /** */
+        private static final long serialVersionUID = 0L;
+
+        /**
+         * Create job with specified argument.
+         *
+         * @param arg   Job argument.
+         * @param debug Flag indicating whether debug information should be printed into node log.
+         */
+        protected VisorMetricJob(@Nullable VisorMetricTaskArg arg, boolean debug) {
+            super(arg, debug);
+        }
+
+        /** {@inheritDoc} */
+        @Override protected Map<String, ?> run(@Nullable VisorMetricTaskArg arg) throws IgniteException {
+            String name = arg.name();
+
+            GridMetricManager mMgr = ignite.context().metric();
+
+            for (ReadOnlyMetricRegistry mReg : mMgr) {
+                String mRegName = mReg.name();
+
+                if (mRegName.equals(name)) {
+                    Map<String, Object> res = new HashMap<>();
+
+                    mReg.forEach(metric -> res.put(metric.name(), valueOf(metric)));
+
+                    return res;
+                }
+
+                String mRegPrefix = mRegName + SEPARATOR;
+
+                if (!name.startsWith(mRegPrefix))
+                    continue;
+
+                if (mRegPrefix.equals(name))
+                    return null;
+
+                String metricName = name.substring(mRegPrefix.length());
+
+                Metric metric = mReg.findMetric(metricName);
+
+                if (metric != null)
+                    return Collections.singletonMap(name, valueOf(metric));
+
+                Object val = searchHistogram(metricName, mReg);
+
+                if (val != null)
+                    return Collections.singletonMap(name, val);
+            }
+
+            return null;
+        }
+
+        /**
+         * Obtains value of the metric.
+         *
+         * @param metric Metric which value should be obtained.
+         * @return Value of the metric.
+         */
+        private Object valueOf(Metric metric) {
```

Comment:
Can `metric.getAsString()` be used instead of this method?

Reply:
I haven't used getAsString() method for primitive types to reduce size of response message and avoid unnecessary converting on the server side.